### PR TITLE
refactor + lit: GlobalAveragePoolOp

### PR DESCRIPTION
**Pull request by vasantha-helprack**
* Adopting `clang-format`
* Improve variable naming
* Extract to function
* Add lit tests

### CODE REVIEW

File: src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp [hunk 1]
```diff
@@ -80,49 +54,30 @@ struct ONNXGlobalAveragePoolOpToTorchLowering : public ConversionPattern {
...
-                 << xtt << "\n"
-                 << "\n";
+                 << xTensor << "\n";
 
-    Value atenGlobAvgpool2d =
-        rewriter.create<AtenAdaptiveAvgPool2dOp>(loc, resultTy, xtt, f1v);
+    Value result = rewriter.create<AtenAdaptiveAvgPool2dOp>(
+        loc, resultType, xTensor, rank);
 
     llvm::outs() << "AtenAdaptiveAvgPool2dOp operation creation"
```
[the remark this answers is not in the document]
Comment (vasantha-helprack) [reply src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp]: Done

File: test/mlir/conversion/onnx_to_torch/add/add_singleip.mlir [hunk 2]
```diff
@@ -1,11 +0,0 @@
-//RUN: onnx-mlir --EmitONNXIR --run-torch-pass %s -o=%t >/dev/null && cat %t.onnx.mlir | FileCheck -v %s
-module {
-  func @main_graph(%arg0: tensor<5xf32>) -> tensor<5xf32> attributes {input_names = ["0"], output_names = ["2"]} {
```
Comment (oowekyala) [inline test/mlir/conversion/onnx_to_torch/add/add_singleip.mlir:3]: Why was this deleted?

Comment (vasantha-helprack) [reply test/mlir/conversion/onnx_to_torch/add/add_singleip.mlir]: We brought this back.

File: test/mlir/conversion/onnx_to_torch/globalavg1/globalavg3.onnx.mlir [hunk 3]
```diff
@@ -0,0 +1,20 @@
+//RUN: onnx-mlir --EmitONNXIR --run-torch-pass %s -o=%t >/dev/null && cat %t.onnx.mlir | FileCheck -v %s
+module attributes {llvm.data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", llvm.target_triple = "x86_64-unknown-linux-gnu"}  {
+  func @main_graph(%arg0: tensor<1x2x15x15xf32>) -> tensor<1x7x1x1xf32> attributes {input_names = ["input.1"], output_names = ["9"]} {
+    %0 = "onnx.Constant"() {value = dense<"0xDEADBEEF"> : tensor<8x2x5x5xf32>} : () -> tensor<8x2x5x5xf32>
+    %1 = "onnx.Constant"() {value = dense<[0.0488194712, 0.0165383574, -0.140125141, 0.0660743564, 0.00143751106, -0.034113694, 0.0405743867, 0.136481673]> : tensor<8xf32>} : () -> tensor<8xf32>
+    %2 = "onnx.Conv"(%arg0, %0, %1) {dilations = [1, 1], group = 1 : si64, kernel_shape = [5, 5], onnx_node_name = "Conv_0", pads = [1, 1, 1, 1], strides = [2, 2]} : (tensor<1x2x15x15xf32>, tensor<8x2x5x5xf32>, tensor<8xf32>) -> tensor<1x8x7x7xf32>
+    %3 = "onnx.MaxPoolSingleOut"(%2) {kernel_shape = [2, 2], onnx_node_name = "MaxPool_1", pads = [0, 0, 0, 0], strides = [2, 2]} : (tensor<1x8x7x7xf32>) -> tensor<1x8x3x3xf32>
```
[the remark this answers is not in the document]
Comment (vasantha-helprack) [reply test/mlir/conversion/onnx_to_torch/globalavg1/globalavg3.onnx.mlir]: @BraunPhilipp, removed two tests and renamed the only one. Please check now.

**Review by BraunPhilipp — DISMISSED** [review]

File: src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp [hunk 4]
```diff
@@ -80,49 +54,22 @@ struct ONNXGlobalAveragePoolOpToTorchLowering : public ConversionPattern {
...
-                 << "\n";
+    auto x = globalAveragePool.X();
+    auto resultType =
+        toTorchType(context, globalAveragePool.getResult().getType());
+    auto xTensor = getTorchTensor(x, rewriter, context, loc);
+    auto rank = getRank(x, rewriter, context, loc);
 
-    Value result = atenGlobAvgpool2d;
+    Value result = rewriter.create<AtenAdaptiveAvgPool2dOp>(
+        loc, resultType, xTensor, rank);
```
[the remark this answers is not in the document]
Comment (siboehm) [reply src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp:72]: I created a sample Colab notebook that illustrates this issues: https://colab.research.google.com/drive/1P7tdabLQh87vwoVEzU9-Qivz21V9Ryn4?usp=sharing

I wrote a PyTorch module that, when converted to onnx generates a `GlobalAvgPool` op, but when converted to torch-mlir, generates a `aten.AdaptiveAvgPool2d` op with an output size of 1 (independent of the rank of the input tensor.

**Review by siboehm — CHANGES_REQUESTED** [review]
The test is called `global_avg_pool`, but it doesn't test GlobalAveragePool, but instead ReduceMean. Can you please add a test for the `onnx.GlobalAveragePool` -> `aten.AdaptiveAvgPool2d`, and move the current Test into the ReduceMean folder?

File: src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (vasantha-helprack) [reply src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp:72]: cc @anushree1998

[the remark this answers is not in the document]
Comment (siboehm) [reply src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp:72]: (sorry I had accidentally changed the colab that I sent in the previous link, I've fixed it now. Just in case someone looked at the link and was confused why the torch-mlir wasn't outputting an AdaptiveAvgPool).

[the remark this answers is not in the document]
Comment (vasantha-helprack) [reply src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp]: @siboehm, we tried running it from this branch, `onnx-mlir --EmitONNXIR --debug  onnx_model.onnx`, we are still getting `onnx.ReduceMean`. 

If I run it with `onnx-mlir --EmitONNXIR --debug  onnx_model.onnx --run-torch-pass`, I get `torch.aten.mean`.

Instead of the rank, now we are passing as list: `[1, 1]`.

[the remark this answers is not in the document]
Comment (siboehm) [reply src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp]: Maybe this has changed w/ different versions of Pytorch / ONNX. If the output shape is correct now, then I'm happy with this PR once a test has been added.

[the remark this answers is not in the document]
Comment (anushree1998) [reply src/Conversion/ONNXToTorch/NN/GlobalAveragePool.cpp]: Hi @siboehm,
This is the result of our model. It also emitting 1:
input: "input"
output: "1"
name: "GlobalAveragePool_0"
op_type: "GlobalAveragePool"

File: test/mlir/conversion/onnx_to_torch/global_avg_pool/global_avg_pool.onnx.mlir [hunk 5]
```diff
@@ -0,0 +1,9 @@
+//RUN: onnx-mlir --EmitONNXIR --run-torch-pass %s -o=%t >/dev/null && cat %t.onnx.mlir | FileCheck -v %s
+module attributes {}  {
+  func @main_graph(%arg0: tensor<1x2x5x5xf32>) -> tensor<1x2x1x1xf32> attributes {input_names = ["input"], output_names = ["1"]} {
+//CHECK: %[[DIM:.*]] = torch.constant.int 0
+//CHECK: torch.aten.mean %arg0, %[[DIM]] : !torch.vtensor<[1,2,5,5],f32>, !torch.int -> !torch.vtensor<[1,1,1,5],f32>
+    %0 = "onnx.ReduceMean"(%arg0) {axes = [1, 2]} : (tensor<1x2x5x5xf32>) -> tensor<1x2x1x1xf32>
```
[the remark this answers is not in the document]
Comment (anushree1998) [reply test/mlir/conversion/onnx_to_torch/global_avg_pool/global_avg_pool.onnx.mlir]: Hi @siboehm,
We have resolved this operation.

**Comment by vasantha-helprack** [comment]
> The test is called `global_avg_pool`, but it doesn't test GlobalAveragePool, but instead ReduceMean. Can you please add a test for the `onnx.GlobalAveragePool` -> `aten.AdaptiveAvgPool2d`, and move the current Test into the ReduceMean folder?

@siboehm, we've fixed it. Please check now.

**Comment by siboehm** [comment]
I think I understand the issue now. The `onnx.GlobalAveragePool` is always being canonicalized to a `onnx.ReduceMean` using [this code](https://github.com/ljfitz/onnx-mlir/blob/b8c9d7a0b62765cd2eb2dc56a3458f6a5d3050b3/src/Dialect/ONNX/Rewrite.td#L557-L558). You can see this in action using `--print-ir-after-all`. So this pass never triggers.

I'll discuss internally what to do about this.

**Comment by siboehm** [comment]
We talked about this issue in our daily meeting. Since the GlobalAveragePool conversion implemented in this PR never triggers, there's no need to keep the code around, as we don't need it & cannot test it. Could you delete this code and instead refactor the existing `onnx.ReduceMean` conversion? I already added a test for it, that code seems to be working well already.

**Comment by vasantha-helprack** [comment]
> We talked about this issue in our daily meeting. Since the GlobalAveragePool conversion implemented in this PR never triggers, there's no need to keep the code around, as we don't need it & cannot test it. Could you delete this code and instead refactor the existing `onnx.ReduceMean` conversion? I already added a test for it, that code seems to be working well already.

@siboehm, while refactoring, I saw that one of the attributes `axes` is not being used in the current implementation of `ReduceMean`. The implementation for this operator is incomplete. Also, looks like we agreed that `ReduceMean` will be implemented by Xilinx. Please let me know to proceed here.

**Comment by siboehm** [comment]
I'm happy with the refactoring as it is. Can we add an assert on the axes parameter (literally something like `assert(axes==[2,3])`? Then it fails when we move to some more general case and we can implement those cases ourselves.

**Comment by vasantha-helprack** [comment]
> I'm happy with the refactoring as it is. Can we add an assert on the axes parameter (literally something like `assert(axes==[2,3])`? Then it fails when we move to some more general case and we can implement those cases ourselves.

@siboehm, please check now.

**Review by siboehm — APPROVED** [review]

**Review by BraunPhilipp — APPROVED** [review]
